### PR TITLE
get TimestampTests working again

### PR DESCRIPTION
**Pull request by zkat**
Fixes: https://github.com/NuGet/Home/issues/8935

### CODE REVIEW

File: test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs [hunk 1]
```diff
@@ -81,15 +81,22 @@ public async Task Timestamp_Verify_WithOfflineRevocation_ReturnsCorrectFlagsAndL
...
+                    {
+                        errors.Count().Should().Be(1);
+                        errors.Should().Contain(w => w.Code == NuGetLogCode.NU3028 && w.Message.Contains("An incomplete certificate revocation check occurred."));
+                    }
                     else
                     {
+                        errors.Count().Should().Be(2);
+                        errors.Should().Contain(w => w.Code == NuGetLogCode.NU3028 && w.Message.Contains("The revocation function was unable to check revocation because the revocation server could not be reached."));
                         errors.Should().Contain(w => w.Code == NuGetLogCode.NU3028 && w.Message.Contains("unable to get certificate CRL"));
                     }
```
Comment (dtivel) [inline test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs:93]: Can you reuse `Test.Utility.Signing.SigningTestUtility.Assert*(…)` methods?

Comment (zkat) [reply test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs:93]: Hm. I tried doing this, but the Utility uses `NU3018`, not `NU3028` (as in the results of the tests). Is this pointing to a bug, or is this simply a corner case in this particular call? Do I need to look into this?

Comment (dtivel) [reply test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs:93]: Create an overload which takes a `NuGetLogCode` and refactor common code into a private method?

The goal is to minimize code duplication and maximize code reuse.

Comment (zkat) [reply test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs:93]: Yup! I was just making sure 3028 wasn't a mistake (looking into it, it looks like we have separate codes for timestamps and signatures, so that makes sense).

Comment (dtivel) [reply test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs:93]: Both [`NU3018`](https://github.com/NuGet/NuGet.Client/blob/7968d2c9ed05195d0581a5a17564602a16b9a4db/src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs#L348-L351) and [`NU3028`](https://github.com/NuGet/NuGet.Client/blob/7968d2c9ed05195d0581a5a17564602a16b9a4db/src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs#L398-L401) pertain to chain building failures.  `NU3018` applies to a primary signature while `NU3028` applies to a timestamp signature.  On the surface, I don't see anything wrong here with this test expecting `NU3028`.

Comment (zkat) [reply test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs:93]: thanks, Damon! :)

File: test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs [hunk 2]
```diff
@@ -661,6 +661,11 @@ public static void VerifyByteArrays(byte[] expected, byte[] actual)
         }
 
         public static void AssertOfflineRevocation(IEnumerable<ILogMessage> issues, LogLevel logLevel)
+        {
+            AssertOfflineRevocation(issues, logLevel, NuGetLogCode.NU3018);
+        }
+
+        public static void AssertOfflineRevocation(IEnumerable<ILogMessage> issues, LogLevel logLevel, NuGetLogCode code)
```
Comment (kartheekp-ms) [inline test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs]: I think these 2 methods can be refactored to a single method because the only difference is the message. May be we can add additional parameter whose value is checked against the `issues` collection (or) invoke [`Assert.All`](https://github.com/xunit/assert.xunit/blob/92395f59e3ff47d6d5a1edeced9e99ccfef9fd37/CollectionAsserts.cs#L24)
```cs
public static void AssertOfflineRevocation
public static void AssertRevocationStatusUnknown

Comment (zkat) [reply test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs]: Thanks for the review! I disagree about refactoring, since the methods are already fairly minimal on their own. I think refactoring them further would be a bit much, no? The bulk of their bodies is, after all, just picking that message, as you say.

File: test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs [hunk 3]
```diff
@@ -81,17 +81,11 @@ public async Task Timestamp_Verify_WithOfflineRevocation_ReturnsCorrectFlagsAndL
...
-                    }
-                    else
-                    {
-                        errors.Should().Contain(w => w.Code == NuGetLogCode.NU3028 && w.Message.Contains("unable to get certificate CRL"));
-                    }
+
+                    errors.Count().Should().Be(RuntimeEnvironmentHelper.IsMacOSX ? 1 : 2);
+
+                    SigningTestUtility.AssertOfflineRevocation(errors, LogLevel.Error, NuGetLogCode.NU3028);
+                    SigningTestUtility.AssertRevocationStatusUnknown(errors, LogLevel.Error, NuGetLogCode.NU3028);
```
Comment (heng-liu) [inline test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs]: Do we need to differentiate three platforms? The tests seem failed.

Comment (zkat) [reply test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs]: haha, I messed up and typoed the message. The test should be passing now.

**Comment by nkolev92** [comment]
@zkat does this need rebased on top of the latest feature branch?

**Comment by zkat** [comment]
It does!

File: test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs [hunk 4]
```diff
@@ -81,17 +81,14 @@ public async Task Timestamp_Verify_WithOfflineRevocation_ReturnsCorrectFlagsAndL
...
-                    else
+                    errors.Count().Should().Be(RuntimeEnvironmentHelper.IsMacOSX ? 1 : 2);
+                    SigningTestUtility.AssertRevocationStatusUnknown(errors, LogLevel.Error, NuGetLogCode.NU3028);
+
+                    if (!RuntimeEnvironmentHelper.IsMacOSX)
                     {
-                        errors.Should().Contain(w => w.Code == NuGetLogCode.NU3028 && w.Message.Contains("unable to get certificate CRL"));
-                    }
+                        SigningTestUtility.AssertOfflineRevocationOnlineMode(errors, LogLevel.Error, NuGetLogCode.NU3028);
+                    }
```
Comment (zivkan) [inline test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs]: Personally, I find this easier to understand:
```cs
                    if (!RuntimeEnvironmentHelper.IsMacOSX)
                    {
                        errors.Count().Should().Be(2);
                        SigningTestUtility.AssertRevocationStatusUnknown(errors, LogLevel.Error, NuGetLogCode.NU3028);
                        SigningTestUtility.AssertOfflineRevocationOnlineMode(errors, LogLevel.Error, NuGetLogCode.NU3028);
                    }
                    else
                    {
                        errors.Count().Should().Be(1);
                        SigningTestUtility.AssertRevocationStatusUnknown(errors, LogLevel.Error, NuGetLogCode.NU3028);
                    }
```

In particular, I find the `.Should.Be(!IsMac ? 1 : 2)` line difficult. If you don't want to split the assertions into distinct mac vs others blocks, then at least consider putting the `mac ? 1 : 2` expression in a separate statement.

**Review by zivkan — APPROVED** [review]

File: test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs (hunk 4 above)
Comment (heng-liu) [reply test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/TimestampTests.cs]: Thanks. Fixed.

File: test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs [hunk 5]
```diff
@@ -678,18 +678,28 @@ public static void AssertOfflineRevocation(IEnumerable<ILogMessage> issues, LogL
...
+        }
+
+        public static void AssertOfflineRevocationOfflineMode(IEnumerable<SignatureLog> issues, LogLevel logLevel, NuGetLogCode code)
         {
             Assert.Contains(issues, issue =>
-                issue.Code == NuGetLogCode.Undefined &&
-                issue.Level == LogLevel.Information &&
+                issue.Code == code &&
+                issue.Level == logLevel &&
                 issue.Message.Contains("The revocation function was unable to check revocation because the certificate is not available in the cached certificate revocation list and NUGET_CERT_REVOCATION_MODE environment variable has been set to offline. For more information, visit https://aka.ms/certificateRevocationMode."));
```
Comment (dominoFire) [inline test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs]: Consider adding `[assembly: InternalsVisibleTo("Test.Utility")]` in [NuGet.Packaging\Properties\AssemblyInfo.cs](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/Properties/AssemblyInfo.cs) to avoid hardcoding the string message. That way, you can replace this line with:

```suggestion
                issue.Message.Contains(NuGet.Packaging.Strings.VerifyCertTrustOfflineWhileRevocationModeOffline));
```

File: test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs [hunk 6]
```diff
@@ -678,18 +678,28 @@ public static void AssertOfflineRevocation(IEnumerable<ILogMessage> issues, LogL
...
+            AssertOfflineRevocationOnlineMode(issues, logLevel, NuGetLogCode.NU3018);
+        }
+
+        public static void AssertOfflineRevocationOnlineMode(IEnumerable<SignatureLog> issues, LogLevel logLevel, NuGetLogCode code)
         {
             Assert.Contains(issues, issue =>
-                issue.Code == NuGetLogCode.NU3018 &&
+                issue.Code == code &&
                 issue.Level == logLevel &&
                 issue.Message.Contains("The revocation function was unable to check revocation because the revocation server could not be reached. For more information, visit https://aka.ms/certificateRevocationMode."));
```
Comment (dominoFire) [inline test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs]: Consider adding `[assembly: InternalsVisibleTo("Test.Utility")]` in [NuGet.Packaging\Properties\AssemblyInfo.cs](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/Properties/AssemblyInfo.cs) to avoid hardcoding the string message. That way, you can replace this line with:

```suggestion
                issue.Message.Contains(NuGet.Packaging.Strings.VerifyCertTrustOfflineWhileRevocationModeOnline));
```

**Review by dominoFire — COMMENTED** [review]
Some comments on string constants in tests

File: test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs (hunk 5 above)
Comment (heng-liu) [reply test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs]: Thanks Fernando! That's cool! Good to know about it:)
I have one question, in AssemblyInfo.cs, for` #if SIGNED_BUILD` part, may I know how to find the value for PublicKey of Test.Utility? Or, do we need that? Thanks!

Comment (heng-liu) [reply test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs]: Never mind. I think I've found it :)

Comment (heng-liu) [reply test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs]: Fixed.

File: test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs (hunk 6 above)
Comment (heng-liu) [reply test/TestUtilities/Test.Utility/Signing/SigningTestUtility.cs]: Fixed. Thanks!

**Review by dominoFire — APPROVED** [review]
